lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1356,9 +1355,49 @@ export class UserAgentApplication {
      */
     private getCachedToken(serverAuthenticationRequest: ServerRequestParameters, account: Account): AuthResponse {
         this.logger.verbose("GetCachedToken has been called");
-        let accessTokenCacheItem: AccessTokenCacheItem = null;
         const scopes = serverAuthenticationRequest.scopes;
+        const authority = serverAuthenticationRequest.authority || this.authority;
+        // Id Token will be returned in every acquireTokenSilentCall regardless of response_type
+        const idTokenCacheItem = this.cacheStorage.getIdToken(this.clientId, account ? account.homeAccountIdentifier : null, authority);
+        
+        if (idTokenCacheItem.value) {
+            this.logger.verbose("Matching ID Token found in cache");
+        } else {
+            this.logger.verbose("No matching ID Token found in cache");
+        }
 
+        let authResponse: AuthResponse = null;
+
+        const idTokenValue = idTokenCacheItem.value;
+        const idTokenObj = (idTokenValue) ? new IdToken(idTokenValue.idToken) : null;
+        // Look in cache for access token in case of token or id_token token response_type
+        if (serverAuthenticationRequest.responseType !== ResponseTypes.id_token) {
+            // If getCachedAccessToken returns null, ResponseUtils.setResponseIdToken will also return null
+            authResponse = this.getCachedAccessToken(serverAuthenticationRequest, account, scopes);
+        } else {
+            // Build an ID Token only AuthResponse when response_type = id_token
+            const aState = this.getAccountState(serverAuthenticationRequest.state);
+            authResponse = {
+                uniqueId: "",
+                tenantId: "",
+                tokenType: ServerHashParamKeys.ID_TOKEN,
+                idToken: null,
+                idTokenClaims: null,
+                accessToken: null,
+                scopes: scopes,
+                expiresOn: null,
+                account: account,
+                accountState: aState,
+                fromCache: true
+            };
+
+            this.logger.verbose("Response generated and token set");
+        }
+        
+        return (idTokenObj) ? ResponseUtils.setResponseIdToken(authResponse, idTokenObj) : authResponse;
```

Comment:
We need to assert that what was requested is being returned if it exists and is not expired. Couple scenarios to keep in mind which I do not see covered here:

1. If `id_token` or `id_token_token` is requested and there is no cached `idToken` or it is expired we need to return `null`
2. If token or `id_token_token` is requested and there is no cached `accessToken` or it is expired we need to return `null`

I think it may be better to make the above if/else block a switch on the `responseType` and in each case build the complete response, verify all the parts you need are there and return either response or `null`.

Reply:
I think the latest changes should cover these observations/suggestions, thanks for the help!